core/synch.c

```diff
@@ -1211,8 +1212,9 @@ synch_with_all_threads(thread_synch_state_t desired_synch_state,
     /* FIXME: this should be a do/while loop - then we wouldn't have
      * to initialize all the variables above
      */
-    while ((!all_synched || exiting_thread_count > expect_exiting)
-           && loop_count++ < max_loops) {
+    while (threads_are_stale ||
+           ((!all_synched || exiting_thread_count > expect_exiting)
+            && loop_count++ < max_loops)) {
```

Comment:
This seems unsafe as it masks the max iter: better to have the max iter check be top level